### PR TITLE
Prevent divide-by-zero in ImageAnimator for images with 0 delay between frames

### LINKED ISSUE #55972: Animated Gif - division by zero
### Description
Binding an animated gif to a picture box will cause a division by zero error

### configuration
VS 2022 P2,. Net Core 6 preview 6 on Windows 10 21H1 (x64) 

### Regression?
Worked on. Net 5

### Other information - Stack Trace

CoreCLR Version: 6.0.21.35212
.NET Version: 6.0.0-preview.6.21352.12
Description: The process was terminated due to an unhandled exception.
Exception Info: System.DivideByZeroException: Attempted to divide by zero.
   at System.Math.DivRem(Int64 a, Int64 b, Int64& result) in System.Private.CoreLib.dll:token 0x600046a+0x0
   at System.Drawing.ImageAnimator.ImageInfo.AdvanceAnimationBy(Int64 milliseconds) in System.Drawing.Common.dll:token 0x6000bed+0x31
   at System.Drawing.ImageAnimator.AnimateImages() in System.Drawing.Common.dll:token 0x60002e0+0x3d
   at System.Threading.Thread.StartHelper.Callback(Object state) in System.Private.CoreLib.dll:token 0x600288d+0x0
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state) in System.Private.CoreLib.dll:token 0x6002a2d+0x40
--- End of stack trace from previous location ---
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state) in System.Private.CoreLib.dll:token 0x6002a2d+0x80
   at System.Threading.Thread.StartCallback() in System.Private.CoreLib.dll:token 0x600280f+0xe

### PR DESCRIPTION
**Pull request by jeffhandley**
Fixes #55972

Per the GIF spec, a frame delay can have a value of 0. That was historically used for 2 purposes:

1. Allow the animation to progress as fast as the renderer chose to animate
2. Utilize the GIF spec feature that has animation wait for user input (mouse click, key press)

While the latter of those features is archaic, we still need to handle 0-delay frames. We assign them our animation delay speed to be "1 tick". It's possible the frame could get skipped if a tick takes longer than the tick time, but this gives a uniform speed for GIFs where all frame delays are 0.

To prevent the divide-by-zero:

1. If a frame delay is <= 0, apply the tick speed
2. Ensure the total animation time increases with each frame, guarding against overflows
3. Update the `ShouldAnimate` property to check that the total animation time is > 0

### CODE REVIEW

File: src/libraries/System.Drawing.Common/src/System/Drawing/ImageInfo.cs [hunk 1]
```diff
@@ -127,15 +130,15 @@ public bool FrameDirty
...
             /// </summary>
-            private long TotalAnimationTime => Animated ? _frameEndTimes![_frameCount - 1] : 0;
+            private long TotalAnimationTime => Animated ? _totalAnimationTime : 0;
 
             /// <summary>
             /// Whether animation should progress, respecting the image's animation support
             /// and if there are animation frames or loops remaining.
             /// </summary>
-            private bool ShouldAnimate => Animated ? (_loopCount == 0 || _loop <= _loopCount) : false;
+            private bool ShouldAnimate => TotalAnimationTime > 0 ? (_loopCount == 0 || _loop <= _loopCount) : false;
```
Comment (jeffhandley) [inline src/libraries/System.Drawing.Common/src/System/Drawing/ImageInfo.cs:141]: This changes the behavior at line 160, preventing animation from occurring if the total animation time is <= 0.

File: src/libraries/System.Drawing.Common/src/System/Drawing/ImageAnimator.cs [hunk 2]
```diff
@@ -35,6 +35,8 @@ namespace System.Drawing
     /// </summary>
     public sealed partial class ImageAnimator
     {
+        internal const int AnimationDelayMS = 40;
```
Comment (danmoseley) [inline src/libraries/System.Drawing.Common/src/System/Drawing/ImageAnimator.cs:38]: I had to go back to find out where 40ms comes from -- I was reading the gif specs, etc. Looking at history, it's "some number a bit shorter than 50ms on the assumption that we won't show more than 20 frames a second". I wonder if that's worth a comment.

File: src/libraries/System.Drawing.Common/src/System/Drawing/ImageInfo.cs [hunk 3]
```diff
@@ -66,7 +67,21 @@ public ImageInfo(Image image)
                             }
 
                             // Frame delays are stored in 1/100ths of a second; convert to milliseconds while accumulating
-                            _frameEndTimes[f] = (lastEndTime += (BitConverter.ToInt32(values, i) * 10));
+                            // Per spec, a frame delay can be 0 which is treated as a single animation tick
+                            int delay = BitConverter.ToInt32(values, i) * 10;
+                            lastEndTime += delay > 0 ? delay : ImageAnimator.AnimationDelayMS;
+
+                            // Guard against overflows
+                            if (lastEndTime < _totalAnimationTime)
```
Comment (danmoseley) [inline src/libraries/System.Drawing.Common/src/System/Drawing/ImageInfo.cs:75]: this is for cases where delay is negative? does that ever happen/is it legal or are you just being careful?

Comment (danmoseley) [reply src/libraries/System.Drawing.Common/src/System/Drawing/ImageInfo.cs:75]: if just being careful, could you instead simply force delay to be no less than zero? then I think you wouldn't need _totalAnimationTime?

File: src/libraries/System.Drawing.Common/src/System/Drawing/ImageAnimator.cs (hunk 2 above)
Comment (jeffhandley) [reply src/libraries/System.Drawing.Common/src/System/Drawing/ImageAnimator.cs:38]: Yep; good call. Sorry about the inconvenience there.

File: src/libraries/System.Drawing.Common/src/System/Drawing/ImageInfo.cs (hunk 3 above)
Comment (jeffhandley) [reply src/libraries/System.Drawing.Common/src/System/Drawing/ImageInfo.cs:75]: I'm being careful mostly. Here's how it comes together though.

1. If the delay is <= 0, then apply the default animation delay
2. If the accumulated last end time overflows, then collapse all of the remaining frames into that instant (essentially skipping any frames beyond the overflow of animation time)
3. Ensure that TotalAnimationTime is a non-zero value if any frames are to be animated
4. Eliminate the array access from TotalAnimationTime, replacing that with _totalAnimationTime

File: src/libraries/System.Drawing.Common/src/System/Drawing/ImageAnimator.cs (hunk 2 above)
Comment (jeffhandley) [reply src/libraries/System.Drawing.Common/src/System/Drawing/ImageAnimator.cs:38]:
```suggestion
        // We use a timer to apply an animation tick speeds of something a bit shorter than 50ms
        // such that if the requested frame rate is about 20 frames per second, we will rarely skip
        // a frame entirely. Sometimes we'll show a few more frames if available, but we will never
        // show more than 25 frames a second and that's OK.
        internal const int AnimationDelayMS = 40;
```

**Comment by AraHaan** [comment]
Speaking of animated images, I wonder when System.Drawing.Common's ImageAnimator can pick up png images with multiple frames as animated?

I guess I could try to see if I can propose an api to help provide support for animated png's if possible on the .NET side of things if that is required then somehow get the image animator to actually check for it all.

**Review by danmoseley — APPROVED** [review]

**Comment by danmoseley** [comment]
Unrelated failures.

**Review by safern — COMMENTED** [review]
LGTM, thanks for fixing so quickly!

**Pull request merged**